File: medium/11. Valudate BST/valudate_bst.go
package main

import "math"

type BST struct {
	Value int

	Left  *BST
	Right *BST
}

func (tree *BST) ValidateBst() bool {
	return _validateBST(tree, math.MinInt32, math.MaxInt32)
}

func _validateBST(tree *BST, min_val, max_val int) bool {
	if tree.Value < min_val || tree.Value >= max_val {
		return false
	}

	if tree.Left != nil && !_validateBST(tree.Left, min_val, tree.Value) {
		return false
	}

	if tree.Right != nil && !_validateBST(tree.Right, tree.Value, max_val) {
		return false
	}

	return true
}

File: medium/11. Valudate BST/valudate_bst.cpp
#include <iostream>

class BST {
public:
  int value;
  BST *left;
  BST *right;

  BST(int val);
  BST &insert(int val);
};

// minimum signed int value
int MIN = -2147483648;
int MAX = 2147483647;

bool _validateBst(BST *tree, int min, int max);

bool validateBst(BST *tree) { return _validateBst(tree, MIN, MAX); }

bool _validateBst(BST *tree, int min, int max) {
  if (tree->value < min || tree->value >= max) {
    return false;
  }
  if (tree->left != NULL && !_validateBst(tree->left, min, tree->value)) {
    return false;
  }
  if (tree->right != NULL && !_validateBst(tree->right, tree->value, max)) {
    return false;
  }
  return true;
}
